File: internal/api/handlers_logs.go
package api

import (
	"log/slog"
	"net/http"
	"strconv"
	"time"

	"github.com/Z-M-Huang/openhive/internal/domain"
)

const (
	defaultLogsLimit = 100
	maxLogsLimit     = 1000
)

// GetLogsHandler returns a handler for GET /api/v1/logs.
// Supports query params: level, component, team, agent, task_id, since, until, limit, offset.
func GetLogsHandler(logStore domain.LogStore, logger *slog.Logger) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		opts := domain.LogQueryOpts{}

		// Parse level filter
		if levelStr := q.Get("level"); levelStr != "" {
			level, err := domain.ParseLogLevel(levelStr)
			if err != nil {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "invalid log level")
				return
			}
			opts.Level = &level
		}

		// Parse string filters (these are literal values, no injection risk via ORM)
		opts.Component = q.Get("component")
		opts.TeamName = q.Get("team")
		opts.AgentName = q.Get("agent")
		opts.TaskID = q.Get("task_id")

		// Parse time filters
		if sinceStr := q.Get("since"); sinceStr != "" {
			t, err := time.Parse(time.RFC3339, sinceStr)
			if err != nil {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "since must be RFC3339 timestamp")
				return
			}
			opts.Since = &t
		}
		if untilStr := q.Get("until"); untilStr != "" {
			t, err := time.Parse(time.RFC3339, untilStr)
			if err != nil {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "until must be RFC3339 timestamp")
				return
			}
			opts.Until = &t
		}

		// Parse pagination
		opts.Limit = defaultLogsLimit
		if limitStr := q.Get("limit"); limitStr != "" {
			limit, err := strconv.Atoi(limitStr)
			if err != nil || limit < 1 {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "limit must be a positive integer")
				return
			}
			if limit > maxLogsLimit {
				limit = maxLogsLimit
			}
			opts.Limit = limit
		}
		if offsetStr := q.Get("offset"); offsetStr != "" {
			offset, err := strconv.Atoi(offsetStr)
			if err != nil || offset < 0 {
				Error(w, http.StatusBadRequest, "INVALID_PARAM", "offset must be a non-negative integer")
				return
			}
			opts.Offset = offset
		}

		entries, err := logStore.Query(r.Context(), opts)
		if err != nil {
			logger.Error("failed to query logs", "error", err)
			Error(w, http.StatusInternalServerError, "INTERNAL_ERROR", "failed to query logs")
			return
		}

		// Return empty array rather than null
		if entries == nil {
			entries = []*domain.LogEntry{}
		}

		JSON(w, http.StatusOK, entries)
	}
}

File: backend/src/api/handlers-logs.ts
/**
 * OpenHive Backend - Log Viewer API Handler
 *
 * Implements GET /api/v1/logs with query params:
 * level, component, team, agent, task_id, since, until, limit, offset.
 * Returns empty array instead of null.
 *
 * Note: ajv-formats is not installed, so format:'date-time' is an annotation
 * only. since/until are validated manually in the handler.
 */

import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';

import type { LogStore } from '../domain/interfaces.js';
import type { LogEntry, LogQueryOpts } from '../domain/types.js';
import type { LogLevel } from '../domain/enums.js';
import type { MiddlewareLogger } from './middleware.js';
import { sendError, sendJSON } from './response.js';
import type { FastifyReplyShim } from './response.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_LOGS_LIMIT = 100;
const MAX_LOGS_LIMIT = 1000;

// ---------------------------------------------------------------------------
// Internal request types
// ---------------------------------------------------------------------------

interface GetLogsQuery {
  level?: LogLevel;
  component?: string;
  team?: string;
  agent?: string;
  task_id?: string;
  since?: string;
  until?: string;
  limit?: number;
  offset?: number;
}

// ---------------------------------------------------------------------------
// JSON schema
// ---------------------------------------------------------------------------

/**
 * JSON schema for GET /api/v1/logs querystring.
 * since/until use type:'string' — date-time validation is done in the handler
 * because ajv-formats is not installed (format keyword is annotation-only).
 */
export const GET_LOGS_QUERY_SCHEMA = {
  querystring: {
    type: 'object',
    additionalProperties: false,
    properties: {
      level: {
        type: 'string',
        enum: ['debug', 'info', 'warn', 'error'],
      },
      component: {
        type: 'string',
        maxLength: 128,
        pattern: '^[a-zA-Z0-9._-]+$',
      },
      team: {
        type: 'string',
        pattern: '^[a-z0-9][a-z0-9-]*[a-z0-9]$',
        maxLength: 64,
      },
      agent: {
        type: 'string',
        maxLength: 128,
      },
      task_id: {
        type: 'string',
        pattern: '^[a-f0-9-]+$',
        maxLength: 128,
      },
      since: { type: 'string' },
      until: { type: 'string' },
      limit: {
        type: 'integer',
        minimum: 1,
        maximum: MAX_LOGS_LIMIT,
        default: DEFAULT_LOGS_LIMIT,
      },
      offset: {
        type: 'integer',
        minimum: 0,
        default: 0,
      },
    },
  },
};

// ---------------------------------------------------------------------------
// Route handler
// ---------------------------------------------------------------------------

/**
 * Handler factory for GET /api/v1/logs.
 */
export function getLogsHandler(logStore: LogStore, logger: MiddlewareLogger) {
  return async (
    request: FastifyRequest<{ Querystring: GetLogsQuery }>,
    reply: FastifyReply,
  ): Promise<void> => {
    const { level, component, team, agent, task_id } = request.query;
    const limit = request.query.limit ?? DEFAULT_LOGS_LIMIT;
    const offset = request.query.offset ?? 0;

    const opts: LogQueryOpts = { level, component, team_name: team, agent_name: agent, task_id, limit, offset };

    // Validate and parse since
    if (request.query.since !== undefined) {
      const d = new Date(request.query.since);
      if (isNaN(d.getTime())) {
        sendError(reply as FastifyReplyShim, 400, 'INVALID_PARAM', 'since must be a valid date-time string');
        return;
      }
      opts.since = d;
    }

    // Validate and parse until
    if (request.query.until !== undefined) {
      const d = new Date(request.query.until);
      if (isNaN(d.getTime())) {
        sendError(reply as FastifyReplyShim, 400, 'INVALID_PARAM', 'until must be a valid date-time string');
        return;
      }
      opts.until = d;
    }

    let entries: LogEntry[];
    try {
      entries = await logStore.query(opts);
    } catch (err) {
      logger.error('failed to query logs', err);
      sendError(reply as FastifyReplyShim, 500, 'INTERNAL_ERROR', 'failed to query logs');
      return;
    }

    // Return empty array rather than null
    sendJSON(reply as FastifyReplyShim, 200, entries ?? []);
  };
}

// ---------------------------------------------------------------------------
// Route registration
// ---------------------------------------------------------------------------

/**
 * Registers all log routes on the Fastify instance.
 */
export function registerLogRoutes(
  fastify: FastifyInstance,
  logStore: LogStore,
  logger: MiddlewareLogger,
): void {
  fastify.get('/api/v1/logs', { schema: GET_LOGS_QUERY_SCHEMA }, getLogsHandler(logStore, logger));
}
